ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp

```diff
@@ -122,33 +159,104 @@ size_t TensorLayout::compute_page_size_bytes(const Size& page_size) const {
+        };
+
+        const auto& logical_shard_shape = memory_config_.shard_spec.has_value() ? Size(memory_config_.shard_spec.value().shape) : compute_default_logical_shard_shape();
+        const auto& physical_shard_shape = compute_physical_shard_shape(logical_shard_shape);
+
+        auto get_physical_size = [](auto original_size, auto logical_shard_size, auto physical_shard_size, auto alignment) -> uint32_t {
+            if (logical_shard_size == 0) {
+                return 0;
+            }
+            // If we always pad to full shards, then return:
+            // auto num_shards = tt::div_up(original_size, logical_shard_size);
+            // return (uint32_t) physical_shard_size * num_shards;
+
+            // But host physical data is only padded up to nearest alignment (and no padding in between shards)
+            // So last shard is only up to nearest alignment
+            auto num_full_shards = original_size / logical_shard_size;
+            auto last_physical_shard_size = CMAKE_UNIQUE_NAMESPACE::round_up(original_size % logical_shard_size, alignment);
+            return (physical_shard_size * num_full_shards + last_physical_shard_size);
+        };
+
+        auto physical_height = get_physical_size(height, logical_shard_shape.height(), physical_shard_shape.height(), alignment_[-2]);
+        auto physical_width = get_physical_size(width, logical_shard_shape.width(), physical_shard_shape.width(), alignment_[-1]);
+
+        Size size{physical_height, physical_width};
+        return size;
+    } else {
+        const int max_rank = std::max(rank, alignment_rank);
+
+        // Iterate dims in reverse order and ensure alignment
+        // Even tensor of rank 0 or 1 must be aligned (to Tile / Page / Shard)
+        for (int i = -1; i >= -max_rank; --i) {
+            auto& dim = i == -1 ? width : height;
+            if(i >= -rank) {
+                dim *= shape[i];
+            }
+
+            // Align the current dimension if alignment is available
+            if (i >= -alignment_rank) {
+                dim = CMAKE_UNIQUE_NAMESPACE::round_up(dim, alignment_[i]);
+            }
         }
-    }
 
-    Size size{height, width};
-    return size;
+        Size size{height, width};
+        return size;
+    }
 }
 
 Size TensorLayout::compute_page_shape(const Size& physical_size) const {
-    return page_config_.get_page_shape(physical_size, dtype_, memory_config_);
+    std::optional<Size> physical_shard_shape = std::nullopt;
+    if (memory_config_.shard_spec.has_value()) {
+        const auto& shard_spec = memory_config_.shard_spec.value();
+        if (shard_spec.mode == ShardMode::PHYSICAL) {
+            physical_shard_shape = shard_spec.shape;
+        } else if (shard_spec.mode == ShardMode::LOGICAL) {
+            physical_shard_shape = compute_physical_shard_shape(shard_spec.shape);
+        } else {
+           TT_THROW("Unsupported shard mode {} in compute_shard_spec_buffer!", shard_spec.mode);
+        }
```

Comment:
we have werror enabled so I assume if switch statement does not handle all options - it won't compile
consider that instead of a runtime error

Reply:
I also assume that. Can we double-check it and replace with switch?

Reply:
I won't have a test to be able to test for compile error, but i can change to switch